src/qibo/models/dbi/utils.py

```diff
@@ -0,0 +1,159 @@
+            # append in output_dict
+            output_dict[op_name] = SymbolicHamiltonian(tensor_op).dense.matrix
+    return output_dict
+
+
+def str_to_op(name: str):
+    tensor_op = 1
+    for qubit, char in enumerate(name):
+        tensor_op *= getattr(symbols, char)(qubit)
+    return tensor_op
+
+
+def select_best_dbr_generator(
+    dbi_object: DoubleBracketIteration,
+    d_list: list,
+    step: float = None,
+    step_min: float = 1e-5,
+    step_max: float = 1,
+    max_evals: int = 200,
+    compare_canonical: bool = True,
+):
+    """Selects the best double bracket rotation generator from a list.
+
+    Args:
+        dbi_object (_DoubleBracketIteration): The object intended for double bracket iteration.
+        d_list (list): List of diagonal operators (np.array) to run from.
+        step (float): Fixed iteration duration.
+            Defaults to ``None``, uses hyperopt.
+        step_min (float): Minimally allowed iteration duration.
+        step_max (float): Maximally allowed iteration duration.
+        max_evals (int): Maximally allowed number of evaluation in hyperopt.
+        compare_canonical (bool): If `True`, the optimal diagonal operator chosen from "d_list" is compared with the canonical bracket.
+
+    Returns:
+        The index of the optimal diagonal operator and respective step duration.
+    """
+    h_before = deepcopy(dbi_object.h)
+    norms_off_diagonal_restriction = []
+    optimal_steps = []
+    for d in d_list:
+        # prescribed step durations
+        if step is None:
+            step = dbi_object.hyperopt_step(
+                step_min=step_min,
+                step_max=step_max,
+                space=hp.uniform,
+                optimizer=tpe,
+                max_evals=max_evals,
+                d=d,
+            )
+        dbi_object(step=step, d=d)
+        optimal_steps.append(step)
+        norms_off_diagonal_restriction.append(dbi_object.off_diagonal_norm)
+        dbi_object.h = deepcopy(h_before)
+    # canonical
+    if compare_canonical is True:
+        generator_type = dbi_object.mode
+        dbi_object.mode = DoubleBracketGeneratorType.canonical
+        if step is None:
+            step = dbi_object.hyperopt_step(
+                step_min=step_min,
+                step_max=step_max,
+                space=hp.uniform,
+                optimizer=tpe,
+                max_evals=max_evals,
+            )
+        dbi_object(step=step)
+        optimal_steps.append(step)
+        norms_off_diagonal_restriction.append(dbi_object.off_diagonal_norm)
+        # print(f'canonical step {step}, loss {dbi_object.off_diagonal_norm}')
+        dbi_object.h = deepcopy(h_before)
+        dbi_object.mode = generator_type
+    # find best d
+    # print(norms_off_diagonal_restriction)
+    idx_max_loss = norms_off_diagonal_restriction.index(
+        min(norms_off_diagonal_restriction)
+    )
+    step_optimal = optimal_steps[idx_max_loss]
+    return idx_max_loss, step_optimal
```

Comment:
I suggest returning the `dbi_object` with the best mode, `step_optimal` and `norms_off_diagonal_restriction`, in this way the user can run the best dbi autonomously after calling this function, i.e. we can delete `select_best_dbr_generator_and_run`.

Reply:
Please correct me if I understood wrongly... so we deepcopy `dbi_object` in the function and update it by default? To run the double bracket rotation, the user will do `dbi_input, idx_max_loss, step_optimal = select_best_dbr_generator(dbi_input)`?

Reply:
> Please correct me if I understood wrongly... so we deepcopy dbi_object in the function and update it by default?

You don't need to deeply copy `dbi_object`, you could just update its mode.

> To run the double bracket rotation, the user will do dbi_input, idx_max_loss, step_optimal = select_best_dbr_generator(dbi_input)?

I suggested something like `dbi_input, norms_off_diagonal_restriction, step_optimal = select_best_dbr_generator(dbi_input, ...)`. In this way you can delete `select_best_dbr_generator_and_run` and if you want to run the best dbi use `dbi_input(step_optimal)`.

Reply:
Actually to further simplify this function can we consider using `hyperopt` directly? At the end we are doing something similar to `hyperopt_step` with the only difference that we are optimizing over a larger space which in this case consist of `(steps, ds, mode)`. I can come up with a prototype.

Reply:
Thanks a lot Andrea! I have tried the code and made simple modifications that allow us to store the losses and update h etc.. However, I have an issue with it being very unstable, in a sense that the off-diagonal norm does not always decrease.
The function:
```python
def hyperopt_step_mode(
                        ham: hamiltonians.Hamiltonian,
                        d: list,
                        step_min: float = 1e-5,
                        step_max: float = 1,
                        max_evals: int = 100,
                        verbose: bool = False):
    # Define the search space
    space = {
        "step": hyperopt.hp.uniform("step", step_min, step_max),
        "mode": hyperopt.hp.choice("mode", [
            {"generator": DoubleBracketGeneratorType.canonical},
            {"generator": DoubleBracketGeneratorType.single_commutator, 
             "d_index": hyperopt.hp.choice("d", list(range(len(d))))}  # Using index for 'd'
        ])
    }

    # Define the loss function
    def loss(args):
        config, step = args["mode"], args["step"]
        dbi = DoubleBracketIteration(hamiltonian=ham, mode=config["generator"])
        if config["generator"] is DoubleBracketGeneratorType.canonical:
            return dbi.loss(step=step)
        else:
            d_index = config["d_index"]
            return dbi.loss(step=step, d=d[d_index])  # Use the index to get the value of 'd'

    # Perform the optimization
    best = hyperopt.fmin(fn=loss,
                         space=space,
                         algo=hyperopt.tpe.suggest,
                         max_evals=max_evals,
                         verbose=verbose)

    # Evaluate the best hyperparameters
    best_evaluated = hyperopt.space_eval(space, best)
    best_loss = loss(best_evaluated)
    output = {
        "mode": best_evaluated["mode"]["generator"],
        "step": best_evaluated["step"],
        "d": d[best_evaluated["mode"]["d_index"]] if "d_index" in best_evaluated["mode"] else None,
        "d_index": best_evaluated["mode"]["d_index"] if "d_index" in best_evaluated["mode"] else None,
        "loss": best_loss,
    }

    return output
```
Test code:
```python
NSTEPS = 15
max_evals = 100
step_max = 1
op_names = Z_names + ['Canonical']
# add in initial values for plotting
off_diagonal_norm_history = [dbi.off_diagonal_norm]
steps = [0]
ham = H_TFIM
dbi_hyper_pauli = DoubleBracketIteration(ham, mode=DoubleBracketGeneratorType.single_commutator)
for _ in range(NSTEPS):
    output = hyperopt_step_mode(dbi_hyper_pauli.h, Z_ops)
    if output['mode'] == DoubleBracketGeneratorType.canonical:
        op_ind = -1
        dbi_hyper_pauli(output['step'], mode=DoubleBracketGeneratorType.canonical)
    else:
        op_ind = output['d_index']
        dbi_hyper_pauli(output['step'], d=output['d'], mode=DoubleBracketGeneratorType.single_commutator)
    print(f"New optimized step at iteration {_+1}/{NSTEPS}: {output['step']} with operator {op_names[op_ind]}, loss {dbi_hyper_pauli.off_diagonal_norm}")
```

Reply:
Thanks @Sam-XiaoyueLi. I will have a look. It would be interesting to understand more why `hyperopt` is so unstable

Reply:
Summary of discussion with Sam:
- often the optimization landscape has spikes
- Sam made some observations qualitatively when hyperopt gets stuck or not, it makes a difference if there are several spikey minima of same depth or not
- keep in mind we are using quite few function evaluations ~100
- I think ultimately we will take scheduling formulas from Andrew, get an initial guess for the apex of the parabola and narrow down the hyperopt window
- the hyperopt window for the first DBI step should be quite large, around 0.3 for what I know
- I added to the timeline the task to do some scaling on the apex of the parabola for the first minimum as the number of qubits increases @wrightjandrew @Sam-XiaoyueLi you could work on this together and then later check with @taufiqmurtadho using tenpy
- we might add a scheduling strategy to have the first hyperopt step extensive, with a large window; then subsequently narrow it down